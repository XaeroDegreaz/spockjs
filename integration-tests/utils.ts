import { SpawnSyncOptions } from 'child_process';
import { sync as run } from 'cross-spawn';
import { getInstalledPathSync } from 'get-installed-path';
import which from 'npm-which';
import * as path from 'path';

/*
  Integration tests execute the test runners with ts-node
  so that we can use the plugin's index.ts directly for
  the babel transformation performed by the test runners.
  This disposes the need to build the plugin separately
  before running the integration tests.

  The main Jest process run by `yarn test` does the same
  thing, it runs in ts-node as well.

  Either way, runners needs to be started with cache disabled,
  because the transform cache is only invalidated when the
  test files that use assertion blocks etc. change,
  but not when the babel plugin that transforms them changes.
*/

// paths

export const resolvePath: (...segments: string[]) => string = path.resolve.bind(
  path,
  __dirname,
);
export const modulePath = (packageName: string) =>
  getInstalledPathSync(packageName, { local: true });

const tsconfigPath = resolvePath('..', 'tsconfig.json');

const tsNodePath = which(__dirname).sync('ts-node');

// babel / typescript helpers

/**
 * Preferred way to use JIT TypeScript compilation,
 * available when @babel/register is not required.
 * No further arguments or environment overrides required.
 * If JIT Babel compilation is also required,
 * fall back to the JitArgs/JitEnv this file exports as well.
 */
export const runWithTypescriptJit = (
  args: string[],
  options?: SpawnSyncOptions,
) => run(tsNodePath, ['--project', tsconfigPath, ...args], options);

export const requireBabelJitArgs = ['--require', '@babel/register'];
export const requireTypescriptJitArgs = ['--require', 'ts-node/register'];

export const babelJitEnv = { BABEL_DISABLE_CACHE: '1' };
export const typescriptJitEnv = { TS_NODE_PROJECT: tsconfigPath };
