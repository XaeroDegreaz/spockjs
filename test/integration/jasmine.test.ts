import { sync as run } from 'cross-spawn';

import { env } from 'process';

import { nodeModulesPath, resolvePath, tsconfigPath } from './utils';

const jasmineCli = resolvePath(nodeModulesPath, 'jasmine', 'bin', 'jasmine');
const cwd = resolvePath('jasmine');

test('produces correct output', () => {
  const { status, stdout } = run('node', [jasmineCli], {
    cwd,
    env: {
      ...env,
      JASMINE_CONFIG_PATH: 'jasmine.json',
      TS_NODE_PROJECT: tsconfigPath,
      BABEL_DISABLE_CACHE: '1',
    },
  });

  expect: status === 1;
  expect(stdout.toString().replace(/Stack:.*/s, '')).toMatchSnapshot();
});
