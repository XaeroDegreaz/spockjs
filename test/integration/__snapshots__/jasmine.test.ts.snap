// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`produces correct output 1`] = `
"Started
.F

Failures:
1) jasmine with spock failing test
  Message:
    AssertionError [ERR_ASSERTION]:   # jasmine.js:11
      
      _assert(x === double(2))
              | |   |         
              | |   4         
              3 false         
      
      [number] double(2)
      => 4
      [number] x
      => 3
      
  "
`;
